R/images.R

```diff
@@ -0,0 +1,204 @@
+#' Capture a plot as a saved file
+#'
+#' Easily generates a .png file (or other graphics file) from a plotting
+#' expression.
+#'
+#' @param expr A plotting expression that generates a plot (or yields an object
+#'   that generates a plot when printed, like a ggplot2). We evaluate this
+#'   expression after activating the graphics device (\code{device}).
+#' @param filename The output filename. By default, a temp file with \code{.png}
+#'   extension will be used; you should provide a filename with a different
+#'   extension if you provide a non-PNG graphics device function.
+#' @param device A graphics device function; by default, this will be either
+#'   \code{\link[grDevices:png]{grDevices::png()}},
+#'   \code{\link[ragg:agg_png]{ragg::agg_png()}}, or
+#'   \code{\link[Cairo:CairoPNG]{Cairo::CairoPNG()}}, depending on
+#'   your system and configuration. See \code{\link{defaultPngDevice}}.
+#' @param width,height,res,... Additional arguments to the \code{device} function.
+#'
+#' @seealso \code{\link{plotTag}} saves plots as a self-contained \code{<img>}
+#'   tag.
+#'
+#' @examples
+#'
+#' # Default settings
+#' res <- capturePlot(plot(cars))
+#'
+#' if (interactive()) browseURL(res)
+#'
+#' # Use custom width/height
+#' pngpath <- tempfile(fileext = ".png")
+#' capturePlot(plot(pressure), pngpath, width = 800, height = 375)
+#'
+#' if (interactive()) browseURL(pngpath)
+#'
+#' # Use SVG
+#' svgpath <- capturePlot(
+#'   plot(pressure),
+#'   tempfile(fileext = ".svg"),
+#'   grDevices::svg,
+#'   width = 8, height = 3.75)
+#'
+#' if (interactive()) browseURL(svgpath)
+#'
+#' # Clean up
+#' unlink(res)
+#' unlink(pngpath)
+#' unlink(svgpath)
+#'
+#' @export
+capturePlot <- function(expr, filename = tempfile(fileext = ".png"),
+  device = defaultPngDevice(), width = 400, height = 400, res = 72,
+  ...) {
+  if (!is.function(device)) {
+    stop(call. = FALSE, "The `device` argument should be a function, e.g. `grDevices::png`")
+  }
+
+  expr <- rlang::enquo(expr)
+
+  tempFile <- missing(filename)
+
+  args <- rlang::list2(width = width, height = height, res = res)
+  argnms <- names(formals(device))
+  if (!"..." %in% argnms) {
+    # Only include `width`, `height`, and `res` if the corresponding formal
+    # parameters are present.
+    args <- args[names(args) %in% argnms | names(args) == "filename"]
+  }
+  args <- c(list(filename = filename), args, rlang::list2(...))
+
+  do.call(device, args)
+  dev <- grDevices::dev.cur()
+  on.exit(grDevices::dev.off(dev), add = TRUE, after = FALSE)
+
+  op <- graphics::par(mar = rep(0, 4))
+  # Prevent examples() from prompting
+  grDevices::devAskNewPage(FALSE)
+  tryCatch(graphics::plot.new(), finally = graphics::par(op))
+
+  tryCatch({
+    result <- withVisible(rlang::eval_tidy(expr))
+    if (result$visible) {
+      capture.output(print(result))
```

Comment:
Why is `capture.output()` used here?

Reply:
`capture.output` is just there to prevent anything from going to the console, in case the return value is not actually a ggplot object or similar, just some random visible value.

I'll fix the `result$value` part, thanks!